### PR TITLE
VMLA Dynamic Iota support with Shape dialect work.

### PR DESCRIPTION
**Pull request by rsuderman**
Updated the VMLA HAL module to support Dynamic iota, along with a corresponding range test.

This includes some changes to the IREE Shapex dialect to handle dynamic tests with a corresponding broadcast_to test.

### CODE REVIEW

**Review by benvanik — CHANGES_REQUESTED** [review]
Keeping the files in the same structure makes it easier when hopping across them. iota fits into the same category as broadcast/tile/etc and should live in a consistent place next to those (so following tile in all files, like vmla.imports.mlir, vmlaops.td, op_kernels, vmla_module.cc, etc).

File: iree/hal/vmla/vmla_module.cc [hunk 1]
```diff
@@ -547,6 +553,11 @@ class VMLAModuleState final {
   // VMLA Ops: arithmetic
   //===--------------------------------------------------------------------===//
 
+  IREE_VMLA_NONARY_OP(IotaI8, kernels::Iota, int8_t);
```
Comment (benvanik) [inline iree/hal/vmla/vmla_module.cc:556]: one more file to fix then LGTM :)

Comment (rsuderman) [reply iree/hal/vmla/vmla_module.cc]: Done!

File: iree/hal/vmla/vmla_module.cc [hunk 2]
```diff
@@ -858,6 +869,11 @@ static const vm::NativeFunction<VMLAModuleState> kVMLAModuleFunctions[] = {
     vm::MakeNativeFunction("shr.i16", &VMLAModuleState::ShrI16),
     vm::MakeNativeFunction("shr.i32", &VMLAModuleState::ShrI32),
 
+    vm::MakeNativeFunction("iota.i8", &VMLAModuleState::IotaI8),
```
Comment (benvanik) [inline iree/hal/vmla/vmla_module.cc:872]: so close :P

Comment (rsuderman) [reply iree/hal/vmla/vmla_module.cc]: ..... I swear this was vscode acting weird.

**Review by benvanik — APPROVED** [review]

**Review by silvasean — APPROVED** [review]
Awesome work Rob! This is super clean!

File: integrations/tensorflow/e2e/range_test.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+  def __init__(self):
+    pass
+
+  @tf.function(input_signature=[
+      tf.TensorSpec([], tf.float32),
+      tf.TensorSpec([], tf.float32)
+  ])
+  def range(self, start, stop):
+    delta = np.array(3, dtype=np.float32)
```
[the remark this answers is not in the document]
Comment (rsuderman) [reply integrations/tensorflow/e2e/range_test.py]: Nope, just from the original copied test. Moved out of the function.

File: iree/compiler/Dialect/Shape/IR/ShapeOps.cpp [hunk 4]
```diff
@@ -49,8 +49,11 @@ static LogicalResult verifyTieShapeOp(TieShapeOp op) {
   auto shapedType = op.operand().getType().dyn_cast<ShapedType>();
   auto rsType = op.shape().getType().dyn_cast<RankedShapeType>();
   if (shapedType && shapedType.hasRank() && rsType) {
-    if (!shapedType.getShape().equals(rsType.getAllDims())) {
-      return op.emitOpError("dims must match between tensor and shape");
+    for (auto it : llvm::zip(shapedType.getShape(), rsType.getAllDims())) {
```
[the remark this answers is not in the document]
Comment (rsuderman) [reply iree/compiler/Dialect/Shape/IR/ShapeOps.cpp:52]: Verify compatible shapes didn't appear to work. Added the test though.

**Pull request merged**